File: vault-plugins/flant_iam/tests/tests/lib/feature_flag.mjs
import Faker from "faker"
import { join } from "path"
import { API } from "./api.mjs"
import { stringifyQuery } from "./endpoint_builder.mjs"
export class FeatureFlagEndpointBuilder {
    one(p = {}, q = {}) {
        return join("/feature_flag", p.name) + stringifyQuery(q)
    }

    collection(p = {}, q = {}) {
        return "/feature_flag" + stringifyQuery(q)
    }
}

export class FeatureFlagAPI {
    constructor(client) {
        this.api = new API(client, new FeatureFlagEndpointBuilder())
    }

    async create(payload, opts) {
        const { data: body } = await this.api.create({ payload, opts })
        if (body.data) return body.data.feature_flag
        return body
    }

    async delete(name, opts) {
        const params = { name }
        const { data: body } = await this.api.delete({ params, opts })
        if (body.data) return body.data.feature_flag
        return body
    }

    async list(opts) {
        const query = { list: true }
        const { data: body } = await this.api.list({ query, opts })
        if (body.data) return body.data.names
        return body
    }
}

export function genFeatureFlag(override = {}) {
    return {
        name: Faker.internet.domainWord(),
        ...override,
    }
}

File: vault-plugins/flant_iam/backend/tests/url/featureflag.go
package url

import (
	"net/url"
	"path"
)

type FeatureFlagEndpointBuilder struct{}

func (b *FeatureFlagEndpointBuilder) OneCreate(_ Params, query url.Values) string {
	return "feature_flag"
}

func (b *FeatureFlagEndpointBuilder) One(params Params, query url.Values) string {
	return path.Join("feature_flag", params["name"].(string)) + "?" + query.Encode()
}

func (b *FeatureFlagEndpointBuilder) Collection(_ Params, query url.Values) string {
	return path.Join("feature_flag") + "/?" + query.Encode()
}

func (b *FeatureFlagEndpointBuilder) Privileged(_ Params, _ url.Values) string {
	panic("this path is not allowed")
}
